primitives/src/unique_vec.rs

```diff
@@ -0,0 +1,334 @@
+//! Module with [`UniqueVec`] type and related functional.
+
+#[cfg(not(feature = "std"))]
+use alloc::{borrow::ToOwned as _, format, string::String, vec::Vec};
+use core::{
+    borrow::Borrow,
+    fmt::{Debug, Display},
+    ops::Deref,
+};
+
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+
+/// Creates a [`UniqueVec`](unique_vec::UniqueVec) from a list of values.
+///
+/// Works like [`vec!`] macro, but does not accept syntax for repeated values
+/// and might return [`Result`].
+#[macro_export]
+macro_rules! unique_vec {
+    () => {
+        $crate::unique_vec::UniqueVec::new()
+    };
+    ($($x:expr),+ $(,)?) => {
+        $crate::unique_vec::UniqueVec::try_from(vec![$($x),+])
+    };
+}
+
+/// Duplicated value at index {index}
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+#[derive(Debug, displaydoc::Display, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct DuplicationError {
+    index: usize,
+}
+
+/// Wrapper type for [`Vec`] which ensures that all elements are unique.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Encode, IntoSchema)]
+pub struct UniqueVec<T: PartialEq>(Vec<T>);
+
+impl<T: PartialEq> Default for UniqueVec<T> {
+    fn default() -> Self {
+        Self(Vec::default())
+    }
+}
+
+impl<T: PartialEq> UniqueVec<T> {
+    /// Create new [`UniqueVec`].
+    #[must_use]
+    pub const fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Try construct [`UniqueVec`] from iterator.
+    ///
+    /// # Errors
+    ///
+    /// - If there are duplicates in the sequence.
+    pub fn try_from_iter<I>(iter: I) -> Result<Self, DuplicationError>
```

Comment:
you should rather implement `FromIterator` that returns `Result<Self, DuplicationError>`

Reply:
See https://github.com/hyperledger/iroha/pull/3903#discussion_r1335687887

Reply:
`FromIterator` is [implemented](https://doc.rust-lang.org/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA,+E%3E%3E-for-Result%3CV,+E%3E) for `Result` in std, but it's not sufficient in our case

Reply:
Works now after we decided to ignore duplciates